torch/profiler/_pattern_matcher.py

```diff
@@ -0,0 +1,88 @@
+from collections import deque
+import re
+
+from torch.profiler import profile
+
+
+class Pattern:
+
+    def __init__(self, prof: profile):
+        self.prof = prof
+        self.description = "Please specify a description for pattern"
+        assert prof.profiler is not None and prof.profiler.kineto_results is not None
+        self.event_tree = prof.profiler.kineto_results.experimental_event_tree(
+        )
+
+    def report(self, event):
+        msg = f"{event.name()}\n{self.description}\n{source_code_location(event)}"
+        return msg
+
+    def match(self, event):
+        raise NotImplementedError
+
+    def matched_events(self):
+        matched_events = []
+        for event in EventTreeDFS(self.event_tree):
+            if self.match(event):
+                matched_events.append(event)
+        return matched_events
+
+    def root_of(self, event):
+        while event.parent:
+            event = event.parent
+        return event
+
+    def siblings_of(self, event):
+        if event.parent:
+            return event.parent.children
+        else:
+            return self.event_tree
+
+    def next_of(self, event):
+        siblings = self.siblings_of(event)
+        index = siblings.index(event)
+        if index + 1 < len(siblings):
+            return siblings[index + 1]
+        else:
+            return None
+
+    def prev_of(self, event):
+        siblings = self.siblings_of(event)
+        index = siblings.index(event)
+        if index - 1 >= 0:
+            return siblings[index - 1]
+        else:
+            return None
+
+
+# Patterns
+
+
+class NamePattern(Pattern):
+
+    def __init__(self, prof: profile, name: str):
+        super().__init__(prof)
+        self.description = f"Matched Name Event: {name}"
+        self.name = name
+
+    def match(self, event):
+        return re.search(self.name, event.name()) is not None
+
+
+def EventTreeDFS(event_tree):
```

Comment:
I really like that you factored the traversal algorithm into a separate function. Any reason why it's not camel case?